File: VSCode_CPP/721.accounts-merge.cpp
/*
 * @lc app=leetcode id=721 lang=cpp
 *
 * [721] Accounts Merge
 */

// @lc code=start
class Solution {
   public:
    unordered_map<string, int> email_to_index;
    int* uf;
    int f(int i) { return uf[i] == i ? i : uf[i] = f(uf[i]); }
    void u(int i, int j) { uf[f(i)] = f(j); }

    vector<vector<string>> accountsMerge(vector<vector<string>>& accounts) {
        int n = accounts.size();
        uf = new int[n];
        for (int i = 0; i < n; i++) {
            uf[i] = i;
            for (int j = 1; j < accounts[i].size(); j++) {
                string& email = accounts[i][j];
                if (email_to_index.count(email))
                    u(email_to_index[email], i);
                else
                    email_to_index[email] = i;
            }
        }

        unordered_map<int, vector<string>> index_to_emails;
        for (auto& email_index_pair : email_to_index) {
            index_to_emails[f(email_index_pair.second)].push_back(
                move(email_index_pair.first));
        }

        vector<vector<string>> result;
        for (auto& index_emails_pair : index_to_emails) {
            result.push_back({});
            result.back().push_back(move(accounts[index_emails_pair.first][0]));
            result.back().insert(result.back().end(),
                                 index_emails_pair.second.begin(),
                                 index_emails_pair.second.end());
            sort(result.back().begin() + 1, result.back().end());
        }
        delete[] uf;
        return result;
    }
};
// @lc code=end

File: Interview/2023/721.accounts-merge.py
#
# @lc app=leetcode id=721 lang=python3
#
# [721] Accounts Merge
#

# @lc code=start
from collections import defaultdict
class Solution:
    def accountsMerge(self, accounts: List[List[str]]) -> List[List[str]]:
        uf = {}
        n = len(accounts)
        for i in range(n):
            uf[i] = i
        
        def find(x):
            px = uf[x]
            while px != uf[px]:
                px = find(px)
            return px
    
        def union(x, y):
            px, py = find(x), find(y)
            if px != py:
                uf[px] = py
            
        email_to_index = {}
        for i in range(n):
            for email in accounts[i][1:]:
                if email in email_to_index:
                    union(i, uf[email_to_index[email]])
                email_to_index[email] = i
        
        root_to_emails = defaultdict(list)
        for i in range(n):
            root_to_emails[find(i)].append(i)
        
        ans = []
        for root, subs in root_to_emails.items():
            name = accounts[root][0]
            emails = set()
            for sub in subs:
                emails.update(accounts[sub][1:])
            ans.append([name] + sorted(emails))
        return ans
            


# @lc code=end

